### PR TITLE
[Preview] Change Feed: Adds Full Fidelity support

### PR DESCRIPTION
**Pull request by ealsur**
## Description

This PR enables the APIs to support Change Feed in Full Fidelity mode. Building on top of the Retention API exposed at #2020 and #2109, once a container has been created with Change Feed retention, Change Feed iterators can be created with Full Fidelity.

Full Fidelity Change Feed exposes creates, replaces, delete operations, including metadata.

The PR exposes a new `ChangeFeedMode` as a **required parameter** to initiate any Change Feed iterator. This prompts the user to make a conscious choice of the mode they want to use.

Current Change Feed (the one available until now) is called `ChangeFeedMode.Incremental`, while the new one is called `ChangeFeedMode.FullFidelity`. 

### Creating a container with retention

```
ContainerProperties containerProperties = new ContainerProperties("myContainer", "/pk");
containerProperties.ChangeFeedPolicy.FullFidelityRetention = TimeSpan.FromMinutes(5);

CosmosContainerResponse containerCreateResponse = await database.CreateContainerAsync(containerProperties, 5000);
```

### Starting an incremental Change Feed (normal Change Feed)

```
FeedIterator feedIterator = container.GetChangeFeedStreamIterator(
        ChangeFeedStartFrom.Now(),
        ChangeFeedMode.Incremental)
```

or

```
FeedIterator<MyType> feedIterator = container.GetChangeFeedIterator<MyType>(
        ChangeFeedStartFrom.Now(),
        ChangeFeedMode.Incremental)
```

### Starting a Change Feed with Full Fidelity

```
FeedIterator feedIterator = container.GetChangeFeedStreamIterator(
        ChangeFeedStartFrom.Now(),
        ChangeFeedMode.FullFidelity)
```

or

```
FeedIterator<MyType> feedIterator = container.GetChangeFeedIterator<MyType>(
        ChangeFeedStartFrom.Now(),
        ChangeFeedMode.FullFidelity)
```

The document payload available during Full Fidelity responses includes the `_metadata` attribute which contains the context information (like `operationType`).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by j82w — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs [hunk 1]
```diff
@@ -1282,6 +1287,7 @@ public abstract class Container
         /// <returns>An iterator to go through the Change Feed.</returns>
         public abstract FeedIterator<T> GetChangeFeedIterator<T>(
             ChangeFeedStartFrom changeFeedStartFrom,
+            ChangeFeedMode changeFeedMode,
```
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs:1290]: I don't mind, it's just that we went over this on the design review and on this GitHub issue: https://github.com/Azure/azure-cosmos-dotnet-v3/issues/2067#issuecomment-751756139

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedMode.cs [hunk 2]
```diff
@@ -10,7 +10,12 @@ namespace Microsoft.Azure.Cosmos
     /// Base class for the change feed mode <see cref="ChangeFeedRequestOptions"/>.
     /// </summary>
     /// <remarks>Use one of the static constructors to generate a ChangeFeedMode option.</remarks>
-    internal abstract class ChangeFeedMode
+#if PREVIEW
+    public
+#else
+    internal
+#endif
+    abstract class ChangeFeedMode
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedMode.cs:18]: I would strongly suggest adding a more explicit comment about the possible risk data loss risk when enabling full fidelity along the lines of this comment in the Java implementation

https://github.com/Azure/azure-sdk-for-java/blob/8401d88b45c40ee704c1aa1abe9fbe5403857290/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosChangeFeedRequestOptions.java#L348-L367

File: Microsoft.Azure.Cosmos/src/Resource/Settings/ChangeFeedPolicy.cs [hunk 3]
```diff
@@ -24,7 +24,12 @@ namespace Microsoft.Azure.Cosmos
     /// </code>
     /// </example>
     /// <seealso cref="ContainerProperties"/>
-    internal sealed class ChangeFeedPolicy
+#if PREVIEW
+    public
+#else
+    internal
+#endif
+    sealed class ChangeFeedPolicy
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/ChangeFeedPolicy.cs:32]: Similar - Java version (comments) https://github.com/Azure/azure-sdk-for-java/blob/8401d88b45c40ee704c1aa1abe9fbe5403857290/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedPolicy.java#L14-L24

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM Matias - thanks!

File: Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedMode.cs (hunk 2 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ChangeFeed/ChangeFeedMode.cs:18]: Added comment

File: Microsoft.Azure.Cosmos/src/Resource/Settings/ChangeFeedPolicy.cs (hunk 3 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/ChangeFeedPolicy.cs:37]: Added comment

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

**Review by j82w — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

**Pull request merged**